snapcraft/internal/meta.py

```diff
@@ -301,50 +301,68 @@ def _wrap_app(self, name, app):
                 raise EnvironmentError(
                     'The specified command {!r} defined in the app {!r} '
                     'does not exist or is not executable'.format(str(e), name))
-        desktop_file = app.pop('desktop', '')
-        if desktop_file:
-            self._reformat_desktop(name, desktop_file)
-
-    def _reformat_desktop(self, name, desktop_file):
-        gui_dir = os.path.join(self.meta_dir, 'gui')
-        desktop_file_path = os.path.join(self._snap_dir, desktop_file)
-        if not os.path.exists(desktop_file_path):
+        desktop_file_name = app.pop('desktop', '')
+        if desktop_file_name:
+            desktop_file = _DesktopFile(name, desktop_file_name,
+                                        self._config_data['name'],
+                                        self._snap_dir, self.meta_dir)
+            desktop_file.parse_and_reformat()
+            desktop_file.write()
+
+
+class _DesktopFile:
+
+    def __init__(self, name, filename, snap_name, snap_dir, meta_dir):
```

Comment:
Last but not least, Can you do
`def __init__(self, *, <rest of args>)`

This will force parametrized calls which makes it easier in the future for future refactoring and reading.